text/0000-type-declarations.md

```diff
@@ -0,0 +1,216 @@
+
+The only Type Declarations supported in WordPress Core at this time, are:
+
+- Class or interface name; e.g., `WP_REST_Request`
+- `self`, which references own class or interface name
+- `array`, to require an array
+
+_These can only be used in function parameters, not as Return Type Declarations._
+
+## When to Use Type Declarations
+
+When you're writing a function or a class method that expects to receive an array or a specific object type, and there is no reason to accept anything other than that specific type.
+
+```php
+protected function foo( Bar $bar, array $items ) {
+  // $bar   must be an instance of the Bar class.
+  // $items must be passed as an array.
+}
+```
+
+## When Not to Use Type Declarations
+
+If the function you're writing is part of a public API and enforcing a specific data type would make the function less flexible in the eyes of a caller. For example, the [`get_post()`](https://developer.wordpress.org/reference/functions/get_post/) function in WordPress core accepts `int|WP_Post|null` as the first parameter. This maximizes flexibility for callers, making the function more convenient in a variety of circumstances.
+
+Likewise, if you're writing a public function for an API and it needs a `WP_Post` instance, it's better not to enforce `WP_Post` with a type declaration. Instead, use the `get_post()` function to resolve the `$post` reference, making it possible for a caller to pass `int|WP_Post|null` to your function as well.
+
+```php
+function foo( $post ) {
+  $post = get_post( $post );
+  ...
+}
+```
+
+# Drawbacks
+
+## Runtime Errors
+
+Using Type Declarations can lead to recoverable fatal errors in PHP 5, and [TypeError](http://php.net/manual/en/class.typeerror.php) exceptions in PHP 7. This is both a blessing and a curse. Runtime errors are mostly advantageous for reasons already stated elsewhere in this RFC; i.e., they help catch bugs right away.
+
+However, unlike [`_doing_it_wrong()`](https://developer.wordpress.org/reference/functions/_doing_it_wrong/), errors associated with invalid data types are more difficult to suppress. Imagine a plugin author writing code that calls upon a core function. If the core function uses Type Declarations to enforce specific data types, and the plugin passes an invalid type, an HTTP error response or WSOD could occur.
+
+# Adoption Strategy
+
+Retrofitting existing functions with Type Declarations is generally discouraged. Doing so would suddenly enforce a rule that did not exist prior, causing runtime errors. For example, imagine themes and plugins calling core functions that previously accepted multiple data types, but now require a specific type. Not good! This scenario should be avoided.
```

Comment:
Retrofitting `public`/`protected` functions/methods should be avoided :100: 

However, what about retrofitting `private` methods ? These are not publicly accessible and should therefore be relatively safe to retrofit.

The only problem there would be plugins which extend a Core class and change the visibility of a `private` method to `protected` or `public`.
In that case, they would receive a `Declaration of Class::method(...) should be compatible with Parent::method(...)` warning. This will not cause WSODs though, so I think encouraging this would not be a bad thing.
See https://3v4l.org/VHoO6

Reply:
> should be avoided 💯

I've updated to this stronger wording you suggested.

---

> However, what about retrofitting private methods ? These are not publicly accessible and should therefore be relatively safe to retrofit.

I agree. Not quite sure how we'd like to word this yet, but giving it some thought. Maybe others have some input they'd like to share on this? It seems reasonably safe to me.

Reply:
@jaswrks I would remove the `:100:` from the RFC text though.....